crates/biome_css_semantic/src/semantic_model/builder.rs

```diff
@@ -53,20 +64,40 @@ impl SemanticModelBuilder {
     pub fn push_event(&mut self, event: SemanticEvent) {
         match event {
             SemanticEvent::RuleStart(range) => {
+                let new_rule_id = self.next_rule_id;
+                self.next_rule_id = RuleId::new(new_rule_id.index() + 1);
+
+                let parent_id = self.current_rule_stack.last().copied();
+
                 let new_rule = Rule {
+                    id: new_rule_id,
                     selectors: Vec::new(),
                     declarations: Vec::new(),
-                    children: Vec::new(),
                     range,
+                    parent_id,
+                    child_ids: Vec::new(),
                 };
-                self.current_rule_stack.push(new_rule);
+
+                if let Some(&parent_id) = self.current_rule_stack.last() {
+                    if let Some(parent_rule) = self.rules_by_id.get_mut(&parent_id) {
+                        parent_rule.child_ids.push(new_rule_id);
+                    }
+                }
+
+                self.rules_by_id.insert(new_rule_id, new_rule.clone());
```

Comment:
```suggestion
                self.rules_by_id.insert(new_rule_id, new_rule);
```

You may can remove `.clone()` here?